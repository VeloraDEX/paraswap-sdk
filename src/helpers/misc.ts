import type { Primitive } from 'ts-essentials';

export const objectToFilledEntries = <T extends Record<string, unknown>>(
  object: T
): [string, string][] => {
  return (
    Object.entries(object)
      // removes keys with undefined values
      .filter(([, value]) => value !== undefined)
      .map(([key, value]) => [key, String(value)])
  );
};

export const constructSearchString = <
  //                               not arrays or mappings
  U extends Record<string, Exclude<Primitive, symbol>>
>(
  queryOptions: U
): `?${string}` | '' => {
  const queryEntries = objectToFilledEntries(queryOptions);

  const queryString = new URLSearchParams(queryEntries).toString();

  // returns empty string or `?${string}`
  return queryString && `?${queryString}`;
};

type MinAxiosError = {
  code?: string;
  request?: any;
  isAxiosError: boolean;
  message: string;
};

type MinAxiosResponse = {
  data: any;
  status: number;
  statusText: string;
  headers: Record<string, any>;
  config: { url?: string; method?: string };
};

type FetcherErrorConstructorInput = MinAxiosError & {
  response?: MinAxiosResponse;
};

export interface FetcherErrorInterface extends FetcherErrorConstructorInput {
  status?: number;
  code?: string;
}

export const isFetcherError = (error: unknown): error is FetcherError => {
  return error instanceof FetcherError;
};

export class FetcherError extends Error implements FetcherErrorInterface {
  code;
  status;
  request;
  response;
  isAxiosError = false;

  public constructor({
    code,
    request,
    response,
    isAxiosError,
    message,
  }: FetcherErrorConstructorInput) {
    super();

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, FetcherError);
    }

    this.name = this.constructor.name;
    this.code = code;
    this.request = request;
    this.isAxiosError = isAxiosError;

    // no response on AxiosError
    if (!response) {
      this.message = message;
      return;
    }

    this.response = response;

    const { data, status } = response;
    this.status = status;
    this.message = isDataWithError(data) ? data.error : message;
  }
}
// to turn `object -> Record` for indexed variable access
function isObject(obj: unknown): obj is Record<string | symbol, any> {
  return !!obj && typeof obj === 'object';
}

export function isDataWithError(data: unknown): data is { error: string } {
  return isObject(data) && typeof data['error'] === 'string';
}

export type ExtractAbiMethodNames<T extends readonly { name: string }[]> =
  T[number]['name'];

// reduce element[] to Object{key: prop, val?: element}
// for example
// gatherObjectsByProp(Token[], token => token.address) => Record<address, Token|undefined>
export function gatherObjectsByProp<T>(
  elements: T[],
  pickProp: (elem: T, index: number) => string
): Record<string, T>;
export function gatherObjectsByProp<T, U>(
  elements: T[],
  pickProp: (elem: T, index: number) => string,
  transfrom: (elem: T, accumElem: U | undefined, index: number) => U
): Record<string, U>;
export function gatherObjectsByProp<T, U>(
  elements: T[],
  pickProp: (elem: T, index: number) => string,
  transform?: (elem: T, accumElem: U | undefined, index: number) => U
): Record<string, T> | Record<string, U> {
  return elements.reduce<Record<string, T> | Record<string, U>>(
    (accum, element, index) => {
      const key = pickProp(element, index);

      const accumElem: T | U | undefined = accum[key];
      const transformedElement = transform
        ? //                       if transform is available, can only be U | undefined
          transform(element, accumElem as U | undefined, index)
        : element;

      accum[key] = transformedElement;

      return accum;
    },
    {}
  );
}

// checks that array has at least one element
export const isFilledArray = <T>(array: T[]): array is [T, ...T[]] => {
  return array.length > 0;
};

export function getRandomInt(): number {
  return Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
}

export function runOnceAndCache<T, Args extends any[]>(
  // can pass `(...any[]) => any but null | undefined`
  func: (...args: Args) => NonNullable<T>
): (...args: Args) => NonNullable<T> {
  let result: NonNullable<T>;

  return (...args) => {
    return result ?? (result = func(...args));
  };
}

export function deriveCompactSignature(signature: string): string {
  // Remove "0x" prefix if present
  if (signature.startsWith('0x')) {
    signature = signature.slice(2);
  }

  // Convert the hex string to a byte array
  const bytes = new Uint8Array(signature.length / 2);
  for (let i = 0; i < signature.length; i += 2) {
    bytes[i / 2] = parseInt(signature.slice(i, i + 2), 16);
  }

  // Validate the signature length (64 or 65 bytes)
  if (bytes.length !== 64 && bytes.length !== 65) {
    throw new Error('Invalid signature length: must be 64 or 65 bytes');
  }

  // Extract r and s components
  const r = `0x${Array.from(bytes.slice(0, 32), (b) =>
    b.toString(16).padStart(2, '0')
  ).join('')}`;
  let v;

  // Handle 64-byte (EIP-2098 compact) and 65-byte signatures
  if (bytes.length === 64) {
    // Extract v from the highest bit of s and clear the bit in s
    v = 27 + (bytes[32]! >> 7);
    bytes[32]! &= 0x7f; // Clear the highest bit
  } else {
    // Extract v directly for 65-byte signature
    v = bytes[64]!;

    // Normalize v to canonical form (27 or 28)
    if (v < 27) {
      v += 27;
    }
  }

  // Compute yParityAndS (_vs) for the compact signature
  const sBytes = Array.from(bytes.slice(32, 64));
  if (v === 28) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    sBytes[0]! |= 0x80; // Set the highest bit if v is 28
  }
  const yParityAndS = `0x${sBytes
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('')}`;

  // Construct the compact signature by concatenating r and yParityAndS
  const compactSignature = r + yParityAndS.slice(2);

  return compactSignature;
}
