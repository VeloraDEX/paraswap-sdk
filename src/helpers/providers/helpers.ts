import { assert } from 'ts-essentials';
import { TypedDataField } from '../../methods/common/orders/buildOrderData';

// regex from @ethersproject/hash TypedDataEncoder.constructor
// may be overly strict, but reliable
const baseTypeRegex = /^([^\x5b]*)(\x5b|$)/;

export function findPrimaryType(
  types: Record<string, TypedDataField[]>
): string {
  const candidates = Object.keys(types);
  const candidatesSet = new Set(candidates);

  candidates.forEach((candidate) => {
    const typedDataFields = types[candidate];
    if (!typedDataFields) return;

    typedDataFields.forEach(({ type }) => {
      // Get the base type (drop any array specifiers)
      const baseType = type.match(baseTypeRegex)?.[1];
      if (!baseType) return;

      // if type was referred to as a child of another type, it can't be the primaryType
      candidatesSet.delete(baseType);
    });
  });

  const [primaryType] = Array.from(candidatesSet);

  assert(
    primaryType,
    `No primary type found in SignableTypedData types, ${JSON.stringify(types)}`
  );

  return primaryType;
}
