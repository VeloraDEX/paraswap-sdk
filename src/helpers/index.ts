import { EthersV5ProviderDeps } from './providers/ethers';
import { EthersV6ProviderDeps } from './providers/ethersV6';
export { constructFetcher as constructAxiosFetcher } from './fetchers/axios';
export { constructFetcher as constructFetchFetcher } from './fetchers/fetch';
export {
  constructEthersV5ContractCaller as constructEthersContractCaller,
  constructEthersV5ContractCaller,
  EthersV5ProviderDeps,
} from './providers/ethers';
export {
  constructContractCaller as constructEthersV6ContractCaller,
  EthersV6ProviderDeps,
} from './providers/ethersV6';
export type EthersProviderDeps = EthersV5ProviderDeps | EthersV6ProviderDeps;

export {
  constructContractCaller as constructWeb3ContractCaller,
  Web3UnpromiEvent,
} from './providers/web3';
export {
  constructContractCaller as constructViemContractCaller,
  MinViemClient,
  txParamsToViemTxParams,
} from './providers/viem';
export { constructToken } from './token';
export { isFetcherError } from './misc';
export type { FetcherError, FetcherErrorInterface } from './misc';
