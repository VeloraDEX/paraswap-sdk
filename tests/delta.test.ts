import * as dotenv from 'dotenv';
import Web3 from 'web3';
import { ethers } from 'ethersV5';
import { ethers as ethersV6 } from 'ethers';
import fetch from 'isomorphic-unfetch';
import {
  constructEthersV5ContractCaller,
  constructEthersV6ContractCaller,
  constructFetchFetcher,
  constructPartialSDK,
  constructWeb3ContractCaller,
  constructGetDeltaContract,
  constructGetDeltaOrders,
  constructGetDeltaPrice,
  constructBuildDeltaOrder,
  constructApproveTokenForDelta,
  constructSignDeltaOrder,
  constructViemContractCaller,
  constructGetPartnerFee,
  SignableDeltaOrderData,
  DeltaPrice,
  constructPostDeltaOrder,
  constructSubmitDeltaOrder,
  PostDeltaOrderParams,
  FetcherFunction,
} from '../src';
import BigNumber from 'bignumber.js';

import erc20abi from './abi/ERC20.json';

import { assert } from 'ts-essentials';
import { HardhatProvider } from './helpers/hardhat';
import { privateKeyToAccount } from 'viem/accounts';
import { createWalletClient, custom, Hex } from 'viem';
import { hardhat } from 'viem/chains';

dotenv.config();

jest.setTimeout(30 * 1000);

const WETH = '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2';
const DAI = '0x6B175474E89094C44Da98b954EedeAC495271d0F';

const chainId = 1;
const srcToken = WETH;
const destToken = DAI;
const srcAmount = (1 * 1e18).toString(); //The source amount multiplied by its decimals

const TEST_MNEMONIC =
  'radar blur cabbage chef fix engine embark joy scheme fiction master release';
//0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9
const wallet = ethers.Wallet.fromMnemonic(TEST_MNEMONIC);
const walletV6 = ethersV6.HDNodeWallet.fromPhrase(TEST_MNEMONIC);

const web3provider = new Web3(HardhatProvider as any);

const ethersProvider = new ethers.providers.Web3Provider(
  HardhatProvider as any
);

const ethersV6Provider = new ethersV6.BrowserProvider(HardhatProvider);
const signerV6 = walletV6.connect(ethersV6Provider);

const fetchFetcher = constructFetchFetcher(fetch);

const signer = wallet.connect(ethersProvider);
const senderAddress = signer.address;

const viemWalletClient = createWalletClient({
  // either walletClient needs to have account set at creation
  // or provider must own the account (for testing can `await viemTestClient.impersonateAccount({ address: senderAddress });`)
  // to be able to sign transactions
  account: privateKeyToAccount(wallet.privateKey as Hex),
  chain: { ...hardhat, id: chainId },
  transport: custom(HardhatProvider),
});

const ethersV5ContractCaller = constructEthersV5ContractCaller(
  {
    ethersProviderOrSigner: signer,
    EthersContract: ethers.Contract,
  },
  senderAddress
);

const ethersV6ContractCaller = constructEthersV6ContractCaller(
  {
    ethersV6ProviderOrSigner: signerV6,
    EthersV6Contract: ethersV6.Contract,
  },
  senderAddress
);

const web3ContractCaller = constructWeb3ContractCaller(
  web3provider,
  senderAddress
);

const viemContractCaller = constructViemContractCaller(
  viemWalletClient,
  senderAddress
);

describe('Delta:methods', () => {
  const deltaSDK = constructPartialSDK(
    {
      chainId: 1,
      fetcher: fetchFetcher,
      contractCaller: ethersV5ContractCaller,
    },
    constructGetDeltaContract,
    constructGetDeltaOrders,
    constructGetDeltaPrice,
    constructBuildDeltaOrder,
    constructApproveTokenForDelta,
    constructGetPartnerFee
  );

  test('Get Delta Price', async () => {
    const deltaPrice = await deltaSDK.getDeltaPrice({
      srcToken: srcToken,
      destToken: destToken,
      amount: srcAmount,
      userAddress: senderAddress,
      srcDecimals: 18,
      destDecimals: 18,
    });

    const staticDeltaPrice: typeof deltaPrice = {
      ...deltaPrice,
      destAmount: 'dynamic_number',
      destAmountBeforeFee: 'dynamic_number',
      srcUSD: 'dynamic_number',
      destUSD: 'dynamic_number',
      destUSDBeforeFee: 'dynamic_number',
      gasCost: 'dynamic_number',
      gasCostBeforeFee: 'dynamic_number',
      gasCostUSD: 'dynamic_number',
      gasCostUSDBeforeFee: 'dynamic_number',
    };

    expect(staticDeltaPrice).toMatchSnapshot();
  });

  test('Get Delta Contract', async () => {
    const deltaContract = await deltaSDK.getDeltaContract();
    expect(deltaContract).toMatchInlineSnapshot(
      `"0x0000000000bbf5c5fd284e657f01bd000933c96d"`
    );
  });

  test('Approve Token For Delta', async () => {
    const deltaContract = await deltaSDK.getDeltaContract();
    assert(deltaContract, 'Delta contract not found');

    const allowanceBefore = await getTokenAllowance({
      tokenAddress: DAI,
      owner: senderAddress,
      spender: deltaContract,
    });

    expect(allowanceBefore.toString()).toEqual('0');

    const amount = '1000000000000000000'; // 1 DAI
    const tx = await deltaSDK.approveTokenForDelta(amount, DAI);
    expect(tx).toBeDefined();
    await tx.wait();

    const allowanceAfter = await getTokenAllowance({
      tokenAddress: DAI,
      owner: senderAddress,
      spender: deltaContract,
    });

    expect(allowanceAfter.toString()).toEqual(amount);
  });

  test('Get Delta Orders for user', async () => {
    const userWithOrders = '0x76176C2971300217E9f48E3dD4e40591500b96Ff';

    const deltaOrders = await deltaSDK.getDeltaOrders({
      userAddress: userWithOrders,
    });

    // Orders that we know the user had in the past
    const staticSliceOfPastOrders = deltaOrders.slice(-2); // first 2 orders historically
    expect(staticSliceOfPastOrders).toMatchSnapshot();
  });

  test('Get Delta Order by Id', async () => {
    const orderId = '50950528-d362-4359-a89e-ed6e49be1a20';
    const deltaOrder = await deltaSDK.getDeltaOrderById(orderId);
    expect(deltaOrder).toMatchSnapshot();
  });

  test('Get PartnerFee', async () => {
    const partnerFee = await deltaSDK.getPartnerFee({ partner: 'paraswap.io' });
    expect(partnerFee).toMatchInlineSnapshot(`
      {
        "partnerAddress": "0x81037e7be71bce9591de0c54bb485ad3e048b8de",
        "partnerFee": 0.15,
        "takeSurplus": false,
      }
    `);
  });

  test('Build Delta Order', async () => {
    const sampleDeltaPrice: DeltaPrice = {
      destAmount: '3163263721766488892666',
      destAmountBeforeFee: '3194635547945152526200',
      destToken: '0x6b175474e89094c44da98b954eedeac495271d0f',
      destUSD: '3166.4269854931',
      destUSDBeforeFee: '3197.8301834931',
      gasCost: '347788',
      gasCostBeforeFee: '124240',
      gasCostUSD: '31.403198',
      gasCostUSDBeforeFee: '11.218137',
      partner: 'anon',
      partnerFee: 0,
      srcAmount: '1000000000000000000',
      srcToken: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
      srcUSD: '3191.5500000000',
    };

    const slippagePercent = 0.5;
    const destAmountAfterSlippage = decreaseBySlippage(
      sampleDeltaPrice.destAmount,
      slippagePercent
    );

    const amount = '1000000000000000000'; // 1 DAI

    const signableOrderData = await deltaSDK.buildDeltaOrder({
      deltaPrice: sampleDeltaPrice,
      owner: senderAddress,
      // beneficiary: anotherAccount, // if need to send destToken to another account
      // permit: "0x1234...", // if signed a Permit1 or Permit2 TransferFrom for DeltaContract
      srcToken: WETH,
      destToken: DAI,
      srcAmount: amount,
      destAmount: destAmountAfterSlippage, // minimum acceptable destAmount
    });

    const staticSignableOrderData: typeof signableOrderData = {
      ...signableOrderData,
      data: {
        ...signableOrderData.data,
        deadline: NaN, // dynamic number
        nonce: 'dynamic_number',
      },
    };
    expect(staticSignableOrderData).toMatchSnapshot();
  });

  let signature = '';

  test.each([
    ['ethersV5', ethersV5ContractCaller],
    ['ethersV6', ethersV6ContractCaller],
    ['web3', web3ContractCaller],
    ['viem', viemContractCaller],
  ])('sign Delta Order with %s', async (libName, contractCaller) => {
    const sdk = constructPartialSDK(
      { chainId: 1, fetcher: fetchFetcher, contractCaller },
      constructSignDeltaOrder
    );

    const sampleOrder: SignableDeltaOrderData = {
      data: {
        beneficiary: '0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9',
        deadline: 1731328853,
        destAmount: '3147447403157656698880',
        destToken: '0x6B175474E89094C44Da98b954EedeAC495271d0F',
        expectedDestAmount: '3163263721766488892666',
        nonce: '1731325253703',
        owner: '0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9',
        partnerAndFee: '0',
        permit: '0x',
        srcAmount: '1000000000000000000',
        srcToken: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
      },
      domain: {
        chainId: 1,
        name: 'Portikus',
        verifyingContract: '0x0000000000bbf5c5fd284e657f01bd000933c96d',
        version: '2.0.0',
      },
      types: {
        Order: [
          {
            name: 'owner',
            type: 'address',
          },
          {
            name: 'beneficiary',
            type: 'address',
          },
          {
            name: 'srcToken',
            type: 'address',
          },
          {
            name: 'destToken',
            type: 'address',
          },
          {
            name: 'srcAmount',
            type: 'uint256',
          },
          {
            name: 'destAmount',
            type: 'uint256',
          },
          {
            name: 'expectedDestAmount',
            type: 'uint256',
          },
          {
            name: 'deadline',
            type: 'uint256',
          },
          {
            name: 'nonce',
            type: 'uint256',
          },
          {
            name: 'partnerAndFee',
            type: 'uint256',
          },
          {
            name: 'permit',
            type: 'bytes',
          },
        ],
      },
    };

    const deltaOrderSignature = await sdk.signDeltaOrder(sampleOrder);
    if (!signature) signature = deltaOrderSignature;
    // signatures match between libraries
    expect(deltaOrderSignature).toEqual(signature);
  });

  const dummyFetcher: FetcherFunction = (params) => {
    // intercept POST requests
    if (params.method === 'POST') {
      return params as any;
    }

    return fetchFetcher(params);
  };

  const mockFetch = jest.fn(dummyFetcher);

  const dummySDK = constructPartialSDK(
    {
      chainId: 1,
      fetcher: mockFetch as FetcherFunction,
      contractCaller: ethersV5ContractCaller,
    },
    constructPostDeltaOrder,
    constructSubmitDeltaOrder
  );

  test('Post Delta Order', async () => {
    const sampleOrderData: SignableDeltaOrderData['data'] = {
      beneficiary: '0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9',
      deadline: NaN, // dynamic number
      destAmount: '3147447403157656698880',
      destToken: '0x6B175474E89094C44Da98b954EedeAC495271d0F',
      expectedDestAmount: '3163263721766488892666',
      nonce: 'dynamic_number',
      owner: '0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9',
      partnerAndFee: '0',
      permit: '0x',
      srcAmount: '1000000000000000000',
      srcToken: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
    };

    const sampleSignature = '0x1234....';

    const input = {
      order: sampleOrderData,
      signature: sampleSignature,
    };

    await dummySDK.postDeltaOrder(input);

    expect(mockFetch).toHaveBeenLastCalledWith({
      data: { ...input, chainId: dummySDK.chainId },
      method: 'POST',
      url: `${dummySDK.apiURL}/delta/orders`,
    });
  });

  test('Submit(=build+sign+post) Delta Order', async () => {
    const sampleDeltaPrice: DeltaPrice = {
      destAmount: '3163263721766488892666',
      destAmountBeforeFee: '3194635547945152526200',
      destToken: '0x6b175474e89094c44da98b954eedeac495271d0f',
      destUSD: '3166.4269854931',
      destUSDBeforeFee: '3197.8301834931',
      gasCost: '347788',
      gasCostBeforeFee: '124240',
      gasCostUSD: '31.403198',
      gasCostUSDBeforeFee: '11.218137',
      partner: 'anon',
      partnerFee: 0,
      srcAmount: '1000000000000000000',
      srcToken: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
      srcUSD: '3191.5500000000',
    };

    const slippagePercent = 0.5;
    const destAmountAfterSlippage = decreaseBySlippage(
      sampleDeltaPrice.destAmount,
      slippagePercent
    );

    const amount = '1000000000000000000'; // 1 DAI

    const input = {
      deltaPrice: sampleDeltaPrice,
      owner: senderAddress,
      // beneficiary: anotherAccount, // if need to send destToken to another account
      // permit: "0x1234...", // if signed a Permit1 or Permit2 TransferFrom for DeltaContract
      srcToken: WETH,
      destToken: DAI,
      srcAmount: amount,
      destAmount: destAmountAfterSlippage, // minimum acceptable destAmount
    };

    await dummySDK.submitDeltaOrder(input);

    const callArgs = mockFetch.mock.lastCall?.[0];
    assert(callArgs, 'No fetch call was made');
    assert('data' in callArgs, 'No data was sent in the fetch call');
    const { order, signature } = callArgs.data as PostDeltaOrderParams;

    expect(signature).toBeDefined();

    const staticSignedOrderData: SignableDeltaOrderData['data'] = {
      ...order,
      deadline: NaN, // dynamic number
      nonce: 'dynamic_number',
    };

    expect(staticSignedOrderData).toMatchSnapshot();
  });
});

function getTokenAllowance({
  tokenAddress,
  owner,
  spender,
}: {
  tokenAddress: string;
  owner: string;
  spender: string;
}): Promise<BigNumber> {
  const contract = new ethers.Contract(tokenAddress, erc20abi, signer);
  return contract.allowance(owner, spender);
}

function decreaseBySlippage(amount: string, slippagePercent: number): string {
  const amountAfterSlippage = BigInt(
    +(+amount * (1 - slippagePercent / 100)).toFixed(0)
  ).toString(10);

  return amountAfterSlippage;
}
