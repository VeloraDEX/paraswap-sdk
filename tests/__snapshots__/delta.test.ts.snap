// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delta:methods Build Delta Order 1`] = `
{
  "data": {
    "beneficiary": "0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9",
    "deadline": NaN,
    "destAmount": "3147447403157656698880",
    "destToken": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
    "expectedDestAmount": "3163263721766488892666",
    "nonce": "dynamic_number",
    "owner": "0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9",
    "partnerAndFee": "0",
    "permit": "0x",
    "srcAmount": "1000000000000000000",
    "srcToken": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
  },
  "domain": {
    "chainId": 1,
    "name": "Portikus",
    "verifyingContract": "0x0000000000bbf5c5fd284e657f01bd000933c96d",
    "version": "2.0.0",
  },
  "types": {
    "Order": [
      {
        "name": "owner",
        "type": "address",
      },
      {
        "name": "beneficiary",
        "type": "address",
      },
      {
        "name": "srcToken",
        "type": "address",
      },
      {
        "name": "destToken",
        "type": "address",
      },
      {
        "name": "srcAmount",
        "type": "uint256",
      },
      {
        "name": "destAmount",
        "type": "uint256",
      },
      {
        "name": "expectedDestAmount",
        "type": "uint256",
      },
      {
        "name": "deadline",
        "type": "uint256",
      },
      {
        "name": "nonce",
        "type": "uint256",
      },
      {
        "name": "partnerAndFee",
        "type": "uint256",
      },
      {
        "name": "permit",
        "type": "bytes",
      },
    ],
  },
}
`;

exports[`Delta:methods Get Delta Order by Id 1`] = `
{
  "chainId": 1,
  "createdAt": "2024-10-18T14:44:03.502Z",
  "deltaVersion": "1.0",
  "expiresAt": "2024-10-18T15:43:16.000Z",
  "id": "50950528-d362-4359-a89e-ed6e49be1a20",
  "order": {
    "beneficiary": "0x76176c2971300217e9f48e3dd4e40591500b96ff",
    "deadline": 1729266196,
    "destAmount": "2950666627548284",
    "destToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
    "nonce": 1729262634617,
    "owner": "0x76176c2971300217e9f48e3dd4e40591500b96ff",
    "permit": "0x",
    "srcAmount": "40000000000000000000",
    "srcToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
  },
  "orderHash": null,
  "partiallyFillable": false,
  "partner": "delta-paraswap.io-local",
  "receivedAmount": null,
  "status": "FAILED",
  "transaction": null,
  "transactions": [],
  "updatedAt": "2024-10-18T14:44:08.895Z",
  "user": "0x76176c2971300217e9f48e3dd4e40591500b96ff",
}
`;

exports[`Delta:methods Get Delta Orders for user 1`] = `
[
  {
    "chainId": 1,
    "createdAt": "2024-10-10T16:18:04.727Z",
    "deltaVersion": "1.0",
    "expiresAt": "2024-10-10T17:17:47.000Z",
    "id": "8515cce6-c7c6-486b-9f1e-5702f204edd6",
    "order": {
      "beneficiary": "0x76176c2971300217e9f48e3dd4e40591500b96ff",
      "deadline": 1728580667,
      "destAmount": "11302885800000000000",
      "destToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
      "nonce": 1728577074265,
      "owner": "0x76176c2971300217e9f48e3dd4e40591500b96ff",
      "permit": "0x00000000000000000000000076176c2971300217e9f48e3dd4e40591500b96ff00000000000000000000000036ff475499e928590659d5b8aa3a34330a583fd900000000000000000000000000000000000000000000000000000000019d278900000000000000000000000000000000000000000000000000000000670bf2ae000000000000000000000000000000000000000000000000000000000000001bf548be9f97f37f0b2ab285bba67c8fdab99c3a08b0fdc0a910267988485535945df93e27958867f1d479c7b7783e98ba586629407e44f8d5c5d4115a7298dca9",
      "srcAmount": "27076489",
      "srcToken": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    },
    "orderHash": null,
    "partiallyFillable": false,
    "partner": "delta-paraswap.io-local",
    "receivedAmount": "17225367867506356154",
    "status": "EXECUTED",
    "transaction": {
      "blobGasPrice": null,
      "blobGasUsed": null,
      "blockHash": "0x21425d0c3625d5e55061ba2fd8ce91621dbee52c7d2846b76dd897d3438a6893",
      "blockNumber": 20936359,
      "contractAddress": null,
      "cumulativeGasUsed": "8881978",
      "from": "0x2e5eF37Ade8afb712B8Be858fEc7389Fe32857e2",
      "gasPrice": "23525770029",
      "gasUsed": "597540",
      "hash": "0x3e5040d187288848ca57e4423d60fd31922b6db7fe636580d96200b03a8a8d8f",
      "index": 57,
      "status": 1,
      "to": "0x1D7405DF25FD2fe80390DA3A696dcFd5120cA9Ce",
    },
    "transactions": [
      {
        "agent": "paraswap",
        "auctionId": "8515cce6-c7c6-486b-9f1e-5702f204edd6",
        "blobGasPrice": 0,
        "blobGasUsed": 0,
        "blockHash": "0x21425d0c3625d5e55061ba2fd8ce91621dbee52c7d2846b76dd897d3438a6893",
        "blockNumber": 20936359,
        "filledPercent": 10000,
        "from": "0x2e5eF37Ade8afb712B8Be858fEc7389Fe32857e2",
        "gasPrice": 23525770029,
        "gasUsed": 597540,
        "hash": "0x3e5040d187288848ca57e4423d60fd31922b6db7fe636580d96200b03a8a8d8f",
        "id": "c729ff7b-ffc8-4c85-88d8-41f4f26668eb",
        "index": 57,
        "partnerFee": "0",
        "protocolFee": "0",
        "receivedAmount": "17225367867506356154",
        "spentAmount": "27076489",
        "status": 1,
        "to": "0x1D7405DF25FD2fe80390DA3A696dcFd5120cA9Ce",
      },
    ],
    "updatedAt": "2024-10-10T16:18:51.447Z",
    "user": "0x76176c2971300217e9f48e3dd4e40591500b96ff",
  },
  {
    "chainId": 1,
    "createdAt": "2024-10-09T16:52:18.826Z",
    "deltaVersion": "1.0",
    "expiresAt": "2024-10-09T17:52:08.000Z",
    "id": "7696f983-4f0d-4bb0-b591-61957abf74de",
    "order": {
      "beneficiary": "0x76176c2971300217e9f48e3dd4e40591500b96ff",
      "deadline": 1728496328,
      "destAmount": "736681085000000000000",
      "destToken": "0xcafe001067cdef266afb7eb5a286dcfd277f3de5",
      "nonce": 1728492729603,
      "owner": "0x76176c2971300217e9f48e3dd4e40591500b96ff",
      "permit": "0x",
      "srcAmount": "21000000000000000",
      "srcToken": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
    },
    "orderHash": null,
    "partiallyFillable": false,
    "partner": "delta-paraswap.io-local",
    "receivedAmount": "1635237633557152096036",
    "status": "EXECUTED",
    "transaction": {
      "blobGasPrice": null,
      "blobGasUsed": null,
      "blockHash": "0x58dbd72e843c1224e8113694b8b4a37e29657f0d52634c0f3f89835a3c7a7187",
      "blockNumber": 20929352,
      "contractAddress": null,
      "cumulativeGasUsed": "19979608",
      "from": "0x2e5eF37Ade8afb712B8Be858fEc7389Fe32857e2",
      "gasPrice": "47757958240",
      "gasUsed": "224182",
      "hash": "0x6d5fa34f4723283dc32496acb8b8faf4bb9e713e3a9d43152ebef7d842c59700",
      "index": 188,
      "status": 1,
      "to": "0x1D7405DF25FD2fe80390DA3A696dcFd5120cA9Ce",
    },
    "transactions": [
      {
        "agent": "paraswap",
        "auctionId": "7696f983-4f0d-4bb0-b591-61957abf74de",
        "blobGasPrice": 0,
        "blobGasUsed": 0,
        "blockHash": "0x58dbd72e843c1224e8113694b8b4a37e29657f0d52634c0f3f89835a3c7a7187",
        "blockNumber": 20929352,
        "filledPercent": 10000,
        "from": "0x2e5eF37Ade8afb712B8Be858fEc7389Fe32857e2",
        "gasPrice": 47757958240,
        "gasUsed": 224182,
        "hash": "0x6d5fa34f4723283dc32496acb8b8faf4bb9e713e3a9d43152ebef7d842c59700",
        "id": "0531cfdf-0732-4757-8a46-3886487eeedd",
        "index": 188,
        "partnerFee": "0",
        "protocolFee": "0",
        "receivedAmount": "1635237633557152096036",
        "spentAmount": "21000000000000000",
        "status": 1,
        "to": "0x1D7405DF25FD2fe80390DA3A696dcFd5120cA9Ce",
      },
    ],
    "updatedAt": "2024-10-09T16:52:37.585Z",
    "user": "0x76176c2971300217e9f48e3dd4e40591500b96ff",
  },
]
`;

exports[`Delta:methods Get Delta Price 1`] = `
{
  "destAmount": "dynamic_number",
  "destAmountBeforeFee": "dynamic_number",
  "destToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
  "destUSD": "dynamic_number",
  "destUSDBeforeFee": "dynamic_number",
  "gasCost": "dynamic_number",
  "gasCostBeforeFee": "dynamic_number",
  "gasCostUSD": "dynamic_number",
  "gasCostUSDBeforeFee": "dynamic_number",
  "partner": "anon",
  "partnerFee": 0,
  "srcAmount": "1000000000000000000",
  "srcToken": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
  "srcUSD": "dynamic_number",
}
`;

exports[`Delta:methods Submit(=build+sign+post) Delta Order 1`] = `
{
  "beneficiary": "0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9",
  "deadline": NaN,
  "destAmount": "3147447403157656698880",
  "destToken": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
  "expectedDestAmount": "3163263721766488892666",
  "nonce": "dynamic_number",
  "owner": "0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9",
  "partnerAndFee": "0",
  "permit": "0x",
  "srcAmount": "1000000000000000000",
  "srcToken": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
}
`;
